runtime/controller/events.go

```diff
@@ -61,12 +61,24 @@ func MakeEvents(mgr ctrl.Manager, controllerName string, ext *events.Recorder) E
 }
 
 // Event emits a Kubernetes event, and forwards the event to the ExternalEventRecorder if configured.
-func (e Events) Event(ctx context.Context, obj client.Object, metadata map[string]string, severity, reason, msg string) {
-	e.Eventf(ctx, obj, metadata, severity, reason, msg)
+// Use MetaEvent or MetaEventf if you want to attach metadata to the external event.
+func (e Events) Event(ctx context.Context, obj client.Object, severity, reason, msg string) {
+	e.MetaEventf(ctx, obj, nil, severity, reason, msg)
 }
 
 // Eventf emits a Kubernetes event, and forwards the event to the ExternalEventRecorder if configured.
-func (e Events) Eventf(ctx context.Context, obj client.Object, metadata map[string]string, severity, reason, msgFmt string, args ...interface{}) {
+// Use MetaEvent or MetaEventf if you want to attach metadata to the external event.
+func (e Events) Eventf(ctx context.Context, obj client.Object, severity, reason, msgFmt string, args ...interface{}) {
+	e.MetaEventf(ctx, obj, nil, severity, reason, msgFmt, args...)
+}
+
+// MetaEvent emits a Kubernetes event, and forwards the event and metadata to the ExternalEventRecorder if configured.
```

Comment:
What kind of things end up in metadata? Calling these `MetaEvent` says to me that it's a different kind of event (one that's on the "meta" level); but I suspect that it's just an event that has extra, caller-supplied fields, is that right?

Reply:
If that latter, may I suggest `EventWithMeta` and `EventWithMetaf` as the method names.

Reply:
We use the metadata to enrich the event with things like source revision, branch, tag, commit. In the future we would like to make the metadata typed, so it can be used to template alerts.

Reply:
I was in doubt about the naming too @squaremo, but your suggestion is a combination of current state of the PR and what I suggested to myself in: https://github.com/fluxcd/pkg/issues/119